blob/conformance.go

```diff
@@ -0,0 +1,460 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package blob provides an easy way to interact with Blob objects within
+// a bucket. It utilizes standard io packages to handle reads and writes.
+package blob
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// bucketMaker describes functions used to create a bucket for a test.
+// Multiple calls to bucketMaker during a test run must refer to the same
+// storage bucket.
+// Functions should return the bucket along with a "done" function to be called
+// when the test is complete.
+// If bucket creation fails, functions should t.Fatal to stop the test.
+type bucketMaker func(t *testing.T) (*Bucket, func())
+
+// RunConformanceTests runs conformance tests for provider implementations
+// of blob.
+func RunConformanceTests(t *testing.T, makeBkt bucketMaker) {
+	tests := []func(*testing.T, bucketMaker){
+		testRead,
+		testWrite,
+		testAttributes,
+		testDelete,
+	}
+	for _, testFn := range tests {
+		testFn(t, makeBkt)
+	}
+}
+
+// testRead tests the functionality of NewReader, NewRangeReader, and Reader.
```

Comment:
Could you write these tests like normal Go tests so that you don't need to have the RunConformanceTests function and utilize `go test`.

Reply:
I don't think so, because these tests don't run as part of "blob", they run as part of each provider-implementation of blob. Currently they run 3x (file, gcs, aws). If I write them as normal tests, they would only run once, for "blob".